resolver-dns/src/main/java/io/netty/resolver/dns/MacOSDnsServerAddressStreamProvider.java

```diff
@@ -0,0 +1,161 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.resolver.dns;
+
+import io.netty.util.NetUtil;
+import io.netty.util.internal.SocketUtils;
+import io.netty.util.internal.UnstableApi;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.InetSocketAddress;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static io.netty.resolver.dns.DefaultDnsServerAddressStreamProvider.DNS_PORT;
+
+/**
+ * Able to parse output of <a href="https://www.unix.com/man-page/osx/8/scutil/">scutil --dns</a> to respect
+ * nameserver configuration.
+ */
+@UnstableApi
+public final class MacOSDnsServerAddressStreamProvider implements DnsServerAddressStreamProvider {
+    private static final Pattern SCUTIL_PATTERN = Pattern.compile("^(.+):(.+)$");
+
+    private static final InternalLogger logger =
+            InternalLoggerFactory.getInstance(MacOSDnsServerAddressStreamProvider.class);
+    private final Map<String, DnsServerAddresses> mappings = buildAddressMap();
+    private final DnsServerAddressStreamProvider fallbackProvider =
+            UnixResolverDnsServerAddressStreamProvider.parseSilently();
+
+    @Override
+    public DnsServerAddressStream nameServerAddressStream(final String hostname) {
+        if (mappings.isEmpty()) {
+            return fallbackProvider.nameServerAddressStream(hostname);
+        }
+
+        String name = hostname;
+        for (;;) {
+            int i = name.indexOf('.', 1);
+            if (i < 0 || i == name.length() - 1) {
+                return fallbackProvider.nameServerAddressStream(hostname);
+            }
+
+            DnsServerAddresses addresses = mappings.get(hostname);
+            if (addresses != null) {
+                return addresses.stream();
+            }
+
+            name = hostname.substring(i + 1);
+        }
+    }
+
+    private static Map<String, DnsServerAddresses> buildAddressMap() {
+        Object result = AccessController.doPrivileged(new PrivilegedAction<Object>() {
+            @Override
+            public Object run() {
+                try {
+                    return Runtime.getRuntime().exec("scutil --dns");
```

Comment:
there was mention of a C API, that would be preferable to exec'ing another process. Can you describe this API, and are there any challenges with using it?

Reply:
The problems would be only in terms of deployment. I guess we could maybe just put in in our native kqueue transport and move on . WDYT ?

Reply:
I don't think we would want to coupled with the kqueue transport. It is still legit to use NIO and want the dns resolver to use the "correct" dns servers. IIUC this would be creating a new submodule like `resolver-dns-macos` with some jni code to call the `dnsinfo.h` methods? We could attempt to bring it in by default here (if supported) otherwise just fallback to `UnixResolverDnsServerAddressStreamProvider`.

Reply:
@Scottmitch sure I can do that... Time to hack some JNI I guess

Reply:
@Scottmitch https://github.com/netty/netty/pull/9161 here you go